src/core/net/dns_client.cpp

```diff
@@ -43,357 +43,777 @@
 
-    // Remove all pending queries.
-    while ((message = mPendingQueries.GetHead()) != nullptr)
+    if (error == OT_ERROR_NONE)
     {
-        queryMetadata.ReadFrom(*message);
-        FinalizeDnsTransaction(*message, queryMetadata, nullptr, 0, OT_ERROR_ABORT);
+        name.SetFromMessage(*mMessage, offset);
+        SuccessOrExit(error = Name::ParseName(*mMessage, offset));
+    }
+    else
+    {
+        VerifyOrExit(error == OT_ERROR_NOT_FOUND);
     }
 
-    return mSocket.Close();
+    SelectSection(aSection, offset, numRecords);
+    SuccessOrExit(error = ResourceRecord::FindRecord(*mMessage, offset, numRecords, aIndex, name, aaaaRecord));
+    aAddress = aaaaRecord.GetAddress();
+    aTtl     = aaaaRecord.GetTtl();
+
+exit:
+    return error;
 }
 
-otError Client::Query(const QueryInfo &aQuery, ResponseHandler aHandler, void *aContext)
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+
+otError Client::Response::FindServiceInfo(Section aSection, const Name &aName, ServiceInfo &aServiceInfo) const
 {
-    otError       error;
-    QueryMetadata queryMetadata;
-    Message *     message     = nullptr;
-    Message *     messageCopy = nullptr;
-    Header        header;
-    Question      question(ResourceRecord::kTypeAaaa);
+    // This method searches for SRV and TXT records in the given
+    // section matching the record name against `aName`, and updates
+    // the `aServiceInfo` accordingly. It also searches for AAAA
+    // record for host name associated with the service (from SRV
+    // record). The search for AAAA record is always performed in
+    // Additional Data section (independent of the value given in
+    // `aSection`).
+
+    otError   error;
+    uint16_t  offset;
+    uint16_t  numRecords;
+    Name      hostName;
+    SrvRecord srvRecord;
+    TxtRecord txtRecord;
+
+    VerifyOrExit(mMessage != nullptr, error = OT_ERROR_NOT_FOUND);
+
+    // Search for a matching SRV record
+    SelectSection(aSection, offset, numRecords);
+    SuccessOrExit(error = ResourceRecord::FindRecord(*mMessage, offset, numRecords, /* aIndex */ 0, aName, srvRecord));
+
+    aServiceInfo.mTtl      = srvRecord.GetTtl();
+    aServiceInfo.mPort     = srvRecord.GetPort();
+    aServiceInfo.mPriority = srvRecord.GetPriority();
+    aServiceInfo.mWeight   = srvRecord.GetWeight();
+
+    hostName.SetFromMessage(*mMessage, offset);
+
+    if (aServiceInfo.mHostNameBuffer != nullptr)
+    {
+        SuccessOrExit(error = srvRecord.ReadTargetHostName(*mMessage, offset, aServiceInfo.mHostNameBuffer,
+                                                           aServiceInfo.mHostNameBufferSize));
+    }
+    else
+    {
+        SuccessOrExit(error = Name::ParseName(*mMessage, offset));
+    }
+
+    // Search in additional section for AAAA record for the host name.
 
-    VerifyOrExit(aQuery.IsValid(), error = OT_ERROR_INVALID_ARGS);
+    error = FindHostAddress(kAdditionalDataSection, hostName, /* aIndex */ 0,
```

Comment:
I think if the server does this, it is not following the spec.

Here is the text from [RFC6763 section-12](https://tools.ietf.org/html/rfc6763#section-12) related to this:
> 
>    DNS has an efficiency feature whereby a DNS server may place
>    additional records in the additional section of the DNS message.
>    These additional records are records that the client did not
>    explicitly request, but the server has reasonable grounds to expect
>    that the client might request them shortly, so including them can
>    save the client from having to issue additional queries.
> 
>    This section recommends which additional records SHOULD be generated
>    to improve network efficiency, for both Unicast and Multicast DNS-SD
>    responses.

----

> DNS client can treat RRs in answer section and additional data section equally? This might even simplify the parsing logic (e.g. no need to select sections).

It would be possible and relatively easy if we allow search to happen in **any** of sections "answer", "authority" and "additional data"(basically over all RRs in the response).

But if we want to just treat "answer" and "additional data" equally (and exclude the "authority"), I don't think it would easy or simple. Note that we have helper methods that search within a section for `index`th occurrence of a record (matching a name and/or record type). If we need to do such iterations and search over two sections (which may not be consecutive in terms of how RRs are encoded in the response message), we need to update the helper methods to enable such behavior and add more complex logic/code.

----

For now, I prefer/suggest to keep the logic as is and follow the RFC spec.